### PR TITLE
Falcon40B TT Implementation

### PR DESCRIPTION
**Pull request by tt-aho**

### CODE REVIEW

**Review by TT-BrianLiu — APPROVED** [review]

File: conftest.py [hunk 1]
```diff
@@ -262,6 +259,29 @@ def device(device_init_destroy):
     ttl.device.DeallocateBuffers(device)
 
 
+@pytest.fixture(scope="function")
+def physical_devices(request):
+    import tt_lib as ttl
+
+    num_devices = ttl.device.GetNumAvailableDevices()
```
Comment (aliuTT) [inline conftest.py:266]: Are you trying to get mmio devices?

can you use `GetNumPCIeDevices`

**Review by mo-tenstorrent — APPROVED** [review]

**Review by TT-billteng — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

File: tt_eager/tt_dnn/op_library/all_gather/kernels/dataflow/interleaved_eth_ring_gather_send.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            curr_idx++;
+            col_idx++;
+            if (col_idx == num_cols) {
+                curr_idx += offset;
+                col_idx = 0;
+            }
+        }
+
+        eth_noc_async_write_barrier();
+        // num_transfers = num_devices - 1
```
Comment (tarafdarTT) [inline tt_eager/tt_dnn/op_library/all_gather/kernels/dataflow/interleaved_eth_ring_gather_send.cpp:76]: remove commented out code

File: tt_eager/tt_dnn/op_library/all_gather/all_gather_op.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    const auto& page_size = input_tensors[0].buffer()->page_size();
+    TT_FATAL(page_size <= MAX_BUFFER, "Page size too large");
+    TT_FATAL(page_size % 32 == 0);
+
+    // TODO: Validate ring
+    TT_FATAL(input_tensors[0].storage_type() == StorageType::DEVICE, "Operands to all_gather need to be on device!");
+    TT_FATAL(input_tensors[0].buffer() != nullptr , "Operands to all_gather need to be allocated in buffers on device!");
+    TT_FATAL(input_tensors[0].memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED);
+    if (this->receiver_device_id == this->sender_device_id) {
+        TT_FATAL(input_tensors[0].device()->get_ethernet_sockets(this->receiver_device_id).size() >= 2, "2 Device ring requires at least 2 eth connections for ring gather");
```
[the remark this answers is not in the document]
Comment (tt-aho) [reply tt_eager/tt_dnn/op_library/all_gather/all_gather_op.cpp:32]: The way this all gather op is implemented is that each tensor goes into its own run operation/op struct, so this validation is being run separately on each input tensor

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp [hunk 4]
```diff
@@ -623,7 +623,7 @@ void Matmul::validate(
 
                         // No padding
                         TT_FATAL(M == per_core_M);
-                        TT_FATAL(N % per_core_N == 0);
+                        // TT_FATAL(N % per_core_N == 0);
```
Comment (tarafdarTT) [inline tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:626]: remove commented out code

File: tt_eager/tt_dnn/op_library/eltwise_binary/eltwise_binary_op.cpp [hunk 5]
```diff
@@ -115,26 +115,31 @@ void EltwiseBinary::validate(const std::vector<Tensor>& input_tensors) const {
...
     TT_FATAL(input_tensor_a.buffer() != nullptr and input_tensor_b.buffer() != nullptr, "Operands to eltwise binary need to be allocated in buffers on device!");
     TT_FATAL((input_tensor_a.layout() == Layout::TILE && input_tensor_b.layout() == Layout::TILE), "Inputs to eltwise binary must be tilized");
+    if (this->in_place) {
+        TT_FATAL(input_tensor_a.memory_config().memory_layout == this->output_mem_config.memory_layout);
+        TT_FATAL(input_tensor_a.memory_config().buffer_type == this->output_mem_config.buffer_type);
+        TT_FATAL(input_tensor_a.dtype() == this->output_dtype);
+    }
     if (input_tensor_a.memory_config().is_sharded()) {
         if (input_tensor_a.memory_config().memory_layout != TensorMemoryLayout::HEIGHT_SHARDED) {
             // If we aren't height sharded, we require all sharding schemes to match until we add blocked reader/writers for width and block sharding
```
Comment (tarafdarTT) [inline tt_eager/tt_dnn/op_library/eltwise_binary/eltwise_binary_op.cpp:125]: do we have an issue tracking the adding of blocked reader/writers?

File: tt_eager/tt_dnn/op_library/all_gather/kernels/dataflow/interleaved_eth_ring_gather_send.cpp (hunk 2 above)
Comment (tt-aho) [reply tt_eager/tt_dnn/op_library/all_gather/kernels/dataflow/interleaved_eth_ring_gather_send.cpp:76]: This is a description of what the value is supposed to be

File: ttnn/ttnn/operations/core.py [hunk 6]
```diff
@@ -466,41 +466,30 @@ def to_memory_config(tensor, memory_config: ttnn.MemoryConfig):
     """
 
     ttl_tensor = tensor.value
+    if tensor.memory_config == memory_config:
```
Comment (tarafdarTT) [inline ttnn/ttnn/operations/core.py:469]: might need to rebase its now ttnn.get_memory_config(tensor)

**Review by ntarafdar — APPROVED** [review]

File: ttnn/ttnn/operations/core.py (hunk 6 above)
Comment (tt-aho) [reply ttnn/ttnn/operations/core.py:469]: Currently cant rebase to main due to some recent changes that require another change to go to main first before we can rebase 🙃 
Will update after

**Review by boris-drazic — APPROVED** [review]

**Review by farbabi — APPROVED** [review]

**Pull request merged**